rust/arrow/src/array/array.rs

```diff
@@ -1540,22 +1540,38 @@ impl<OffsetSize: StringOffsetSizeTrait> GenericStringArray<OffsetSize> {
             offsets.push(length_so_far);
             values.extend_from_slice(s.as_bytes());
         }
-        let array_data = ArrayData::builder(data_type)
+        let array_data = ArrayData::builder(OffsetSize::DATA_TYPE)
             .len(v.len())
             .add_buffer(Buffer::from(offsets.to_byte_slice()))
             .add_buffer(Buffer::from(&values[..]))
             .build();
         Self::from(array_data)
     }
 
-    pub(crate) fn from_opt_vec(v: Vec<Option<&str>>, data_type: DataType) -> Self {
-        let mut offsets = Vec::with_capacity(v.len() + 1);
+    pub(crate) fn from_opt_vec(v: Vec<Option<&str>>) -> Self {
+        let iter = v.iter().map(|e| e.map(|e| e.to_string()));
+        GenericStringArray::from_iter(iter)
+    }
+}
+
+impl<'a, Ptr, OffsetSize: StringOffsetSizeTrait> FromIterator<Ptr>
```

Comment:
I meant type params like:
```suggestion
impl<'a, P, OS: StringOffsetSizeTrait> FromIterator<P>
```

Reply:
I see. I have no strong feelings about it, I just think that explicit is better than implicit, and `P` or `OS` forces the reader to go to the `where` clause to understand what they mean. OTOH, generics are generics, so...

I will leave it like this just because the rest of the code base around offsets is like this.

Reply:
Happy with it as is, if we need to change, we can do that for the whole module/codebase in future